xarray/indexes/multipandasindex.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+from typing import Any, Hashable, Mapping, TypeVar
+
+import numpy as np
+
+from ..core.indexes import Index, IndexVars, PandasIndex
+from ..core.indexing import IndexSelResult, merge_sel_results
+from ..core.utils import Frozen
+from ..core.variable import Variable
+
+T_MultiPandasIndex = TypeVar("T_MultiPandasIndex", bound="MultiPandasIndex")
+
+
+class MultiPandasIndex(Index):
+    """Helper class to implement meta-indexes encapsulating
+    one or more (single) pandas indexes.
+
+    Each pandas index must relate to a separate dimension.
+
+    This class shoudn't be instantiated directly.
+
+    """
+
+    indexes: Frozen[Hashable, PandasIndex]
+    dims: Frozen[Hashable, int]
+
+    __slots__ = ("indexes", "dims")
+
+    def __init__(self, indexes: Mapping[Hashable, PandasIndex]):
```

Comment:
Always better to use Any for Mapping inputs, due to covariance.
Use `Mapping[Any, PandasIndex]`